lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -1048,6 +1053,282 @@ static bool checkOSLogMessageIsConstant(SingleValueInstruction *osLogMessage,
+             "end of scope or ref-count is deleted without its operand");
+      ValueOwnershipKind operandOwnershipKind = operandValue.getOwnershipKind();
+      UseLifetimeConstraint lifetimeConstraint =
+          operand.getOwnershipKindMap().getLifetimeConstraint(
+              operandOwnershipKind);
+      if (lifetimeConstraint == UseLifetimeConstraint::MustBeInvalidated) {
+        // Since deadInst cannot be an end of scope instruction (checked above),
+        // this must be a consuming use of an owned value.
+        assert(operandOwnershipKind == ValueOwnershipKind::Owned);
+        SILBuilderWithScope builder(deadInst);
+        builder.emitDestroyValueOperation(deadInst->getLoc(), operandValue);
+      }
+    }
+  }
+  // Drop all operand references of deadInsts.
+  for (SILInstruction *deadInst : deadInsts) {
+    deadInst->dropAllReferences();
+  }
+}
+
+/// Try to eliminate the OSLogMessage instance \c oslogMessage passed to the os
+/// log call and clean up its dependencies. If the instance cannot be
+/// eliminated, it implies that either the instance is not auto-generated or the
+/// implementation of the os log overlay is incorrect. Therefore emit
+/// diagnostics in such cases.
+static void tryEliminateOSLogMessage(SingleValueInstruction *oslogMessage) {
+  SILFunction *fun = oslogMessage->getFunction();
+  SourceLoc oslogMessageLoc = oslogMessage->getLoc().getSourceLoc();
+
+  // A worklist that contains instructions that are possibly dead.
+  SmallVector<SILInstruction *, 32> worklist;
+  // Set of all dead instructions discovered during the fixpoint iteration.
+  SmallPtrSet<SILInstruction *, 32> allDeadInstructions;
+
+  // Initialize the worklist with oslogMessage and all its transitive users.
+  worklist.push_back(oslogMessage);
+  SmallVector<SILInstruction *, 8> oslogMessageUsers;
+  getTransitiveUsers(oslogMessage, oslogMessageUsers);
+  worklist.append(oslogMessageUsers.begin(), oslogMessageUsers.end());
+
+  // In the following iteration, the worklist could grow in size iff an
+  // instruction is found to be dead. An instruction that is found to be dead
+  // will never be added again to the worklist. So eventually the worklist will
+  // become empty.
+  while (!worklist.empty()) {
+    SILInstruction *candidateInst = worklist.pop_back_val();
+    // Bail out if it is already known to be dead and handled.
+    if (allDeadInstructions.count(candidateInst))
+      continue;
+    if (!isInstructionDead(candidateInst))
+      continue;
+    // If an instruction is dead, it means all its transitive users are also
+    // dead. The transitive users could be writes into memory allocated by
+    // the instruction (if inst is alloc_stack), or could be clean up of the
+    // reference counts e.g. destory_value or destroy_addr instructions.
+    SmallVector<SILInstruction *, 4> deadInsts;
+    for (SILValue result : candidateInst->getResults()) {
+      getTransitiveUsers(result, deadInsts);
```

Comment:
Yes. That is an interesting question. There are two cases as I have mentioned in the comments above when this can happen. (a) A owned SIL value that will be released e.g. struct A(...).  (b) Alloc stack. Let's first consider (a) e.g. 
```
%v =  struct A() 
destroy %v
```
`destroy %v` will not be considered dead by itself. This is because, it doesn't have any uses but it will be considered to have a side effect. However, when looking at `%v = struct A()` that will be considered dead as `%v` is not used in anything but destroys. 

(b) Alloc_stack is the more interesting scenario. 

The code that checks whether an `%v = alloc_stack` is dead has to check all the uses of `%v`. While one of its users say `u` (probably something like `%u = load  %v`) was dead and that is why we added  `%v = alloc_stack` to the queue, other users of it need not be visited before `%v`. Its other uses are checked only to determine whether the `alloc_stack` is really dead. Those uses would not have been seen before. E.g. the SIL test `testDCEOfAllocStackWithConstEvalCalls` in file:`OSLogPrototypeCompileTest.sil` has an example of this scenario. I have provided some details there about this check.

Reply:
Yes. Thanks. This also became more clear by the time I had finished reviewing.